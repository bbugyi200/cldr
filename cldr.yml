### CLDR Configuration File

# URL of this project's github repository.
github_repo: "https://github.com/bbugyi200/cldr"

# Do NOT change the 'current_version' option manually!
current_version: "2.3.1"

# Fake Jira base URL (for tests)... Remove ASAP.
jira_base_url: "https://jira.prod.company.com"

# Update the 'version_part_to_bump' option via `bumper {patch | minor | major}`
# when you want to release a new version of this project.
version_part_to_bump: "none"
