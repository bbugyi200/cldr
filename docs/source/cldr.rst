cldr package
============

.. automodule:: cldr
   :members:
   :undoc-members:
   :show-inheritance:
