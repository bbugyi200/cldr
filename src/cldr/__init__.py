"""ChangeLog Driven Releases (CLDR)...

Command-line tool to make managing a changelog file easier.
"""


__author__ = "Bryan M Bugyi"
__email__ = "bryanbugyi34@gmail.com"
__version__ = "0.2.0"
