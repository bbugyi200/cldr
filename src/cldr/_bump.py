"""Logic for bumping the project version on new releases."""

from typing import Literal


BumpPart = Literal["major", "minor", "patch"]
