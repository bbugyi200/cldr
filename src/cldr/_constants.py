"""Contains constant variables."""

from __future__ import annotations

from typing import Dict, Final, Literal


Kind = Literal["add", "chg", "dep", "fix", "misc", "rm", "sec"]

KIND_TO_SECTION_MAP: Final[Dict[Kind, str]] = {
    "add": "Added",
    "chg": "Changed",
    "dep": "Deprecated",
    "fix": "Fixed",
    "misc": "Miscellaneous",
    "rm": "Removed",
    "sec": "Security",
}

BULLET_EXPLANATION = f"""\
All bullet lines must be of the form `* KIND(TAG_LIST): BODY` or `* KIND:
BODY`, where `KIND` is one of `{sorted(set(KIND_TO_SECTION_MAP))}`, `BODY` is a
sentence or two about the changes you made, `TAG_LIST` is one or more `TAG`s
separated by commas, and `TAG` is one of `'bc'` (to indicate a breaking
change), `GITHUB_ISSUE`, `GITHUB_PR`, `JIRA_ISSUE`, or `RELATIVE_COMMIT`.
These latter tag types are described below:

* `GITHUB_ISSUE`: is either of the form `#N`, `REPO#N`, or `ORG/REPO#N`.
* `GITHUB_PR`: is either of the form `!N`, `REPO!N`, or `ORG/REPO!N`.
* `JIRA_ISSUE`: is either of the form `ORG-N` or `N` (this latter form can only
be used if the 'jira_org' option is set in your project's pyproject.toml).
* `RELATIVE_COMMIT`: is used to reference a commit hash relative to the commit
that added the current changelog bullet. This tag is of the form `cN` where `N`
is an integer offset (e.g. the `c1` tag will reference the commit made directly
before the commit that added the changelog bullet). NOTE: If the current
changelog bullet's `BODY` is `'...'` and the bullet references a commit,
`BODY` will be taken to be the subject of the referenced commit's message.

Here is an example of a valid bullet which references the CSRE-103 Jira issue:
`* add(csre-103): Added the cldr.py script.`
"""

_CLDR_URL = "https://bbgithub.dev.bloomberg.com/ComplianceSRE/tools/blob/master/src/bloomberg/compliance/sre/tools/cldr.py"
UNRELEASED_BEGIN = f"""\
The unreleased section is unique in that we do not add content to it directly.
Instead, developers of this project add specially formatted bullets to files of
the form `{{0}}/USER@BRANCH.md`. Refer to the [{{0}}/README.md] file or the
[cldr] script (which consumes these bullets when a new version of this project
is released) for more information.

[{{0}}/README.md]: {{1}}/tree/master/{{0}}
[cldr]: {_CLDR_URL}
""".format

README_CONTENTS = f"""\
# Changelog Bullet Files

This directory should contain markdown files of the form `USER@BRANCH.md`. Each
of these files should contain one or more bullets of the form described in the
following paragraph. These bullets will be consumed by the [cldr] script when a
new version of this project is released.

{BULLET_EXPLANATION}

[cldr]: {_CLDR_URL}
"""

PROJECT_NAME: Final = "cldr"
