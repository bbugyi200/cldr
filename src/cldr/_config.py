"""Command-line tool to make managing a changelog file easier.

Examples:
    # Prints new changelog file contents to STDOUT.
    cldr build -V 1.2.3

    # Updates CHANGELOG.md in-place.
    cldr build -V 1.2.3 -i

    # Add a new bullet to the 'Added' section of the next release's
    # CHANGELOG.md section.
    cldr add "Added new feature."

    # Add a new bullet to the 'Added' section of the next release's
    # CHANGELOG.md section, which is related to the CSRE-123 jira issue.
    cldr add -t csre-123 "Added new feature."

    # Add a new bullet to the 'Fixed' section...
    cldr fix "Fixed a bug."

    # Add a new bullet to the 'Miscellaneous' section...
    cldr misc "Did something unreleated to any feature."

    # Add a new bullet to the 'Changed' section...
    cldr mod "Changed an existing feature."

    # Add a new bullet to the 'Removed' section...
    cldr rm "Removed an existing feature."

    # Print internal cldr information to STDOUT as JSON data.
    cldr info
"""

# NOTE: The above docstring is used by clack for the command-line --help
#   message. This module is used to define clack configuration classes and the
#   clack parser function.
from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import Any, Dict, List, Literal, Optional, Sequence

import clack
from eris import ErisError, Err, Ok, Result
import toml
from typist import literal_to_list

from ._constants import PROJECT_NAME, Kind


BUildCommand = Literal["build"]
InfoCommand = Literal["info"]
NewCommand = Literal["new"]
Command = Literal[
    BUildCommand, InfoCommand, NewCommand
]  # available CLI sub-commands


class Config(clack.Config):
    """TODO"""

    command: Command

    # --- Options
    changelog_dir: Path = Path("changelog")


class BuildConfig(Config):
    """TODO"""

    command: BUildCommand

    # --- OPTIONS
    changelog: Path = Path("CHANGELOG.md")
    in_place: bool = False
    new_version: str


class NewConfig(Config):
    """TODO"""

    command: NewCommand

    # --- ARGS
    kind: Kind

    # --- OPTIONS
    body: Optional[str] = None
    bullet_file_name: Optional[str] = None
    commit_changes: bool = True
    tags: Optional[List[str]] = None


class InfoConfig(Config):
    """TODO"""

    command: InfoCommand


@lru_cache
def github_repo() -> str:
    """TODO"""
    conf = _get_conf().unwrap()
    result: Optional[str] = conf.get("github_repo")
    assert result is not None
    return result


@lru_cache
def jira_org() -> Optional[str]:
    """TODO"""
    conf = _get_conf().unwrap()

    result: Optional[str] = conf.get("jira_org")
    if result is None:
        return None
    else:
        return result.upper()


@lru_cache
def _get_conf() -> Result[Dict[str, Any], ErisError]:
    def error(emsg: str) -> Err[Any, ErisError]:
        return Err(
            "{}\n\nIn order to use the 'cldr' script, this project's"
            " pyproject.toml file must have a [tool.cldr] section that defines"
            " a 'github_repo' option and (optionally) a 'jira_org' option."
            .format(emsg)
        )

    pyproject_toml = Path("pyproject.toml")
    if pyproject_toml.exists():
        conf = toml.loads(pyproject_toml.read_text())
    else:
        return error("The pyproject.toml file does not exist.")

    result = conf.get("tool", {}).get(PROJECT_NAME)
    if result is None:
        return error(
            f"The pyproject.toml file does not contain a [tool.{PROJECT_NAME}]"
            " section."
        )

    if result.get("github_repo") is None:
        return error(
            "The [tool.cldr] section in the pyproject.toml file does not set"
            " the 'github_repo' option."
        )

    return Ok(result)


def clack_parser(argv: Sequence[str]) -> dict[str, Any]:
    """TODO"""
    parser = clack.Parser()
    parser.add_argument(
        "--changelog-dir",
        type=Path,
        help=(
            "Can be used to explicitly specify the path of the changelog"
            " directory where unreleased change bullets are stored."
            " Defaults to '%(default)s'."
        ),
    )

    new_command = clack.new_command_factory(parser)

    build_parser = new_command(
        "build",
        help=(
            "Use the bullets found in the changelog directory to generate"
            " a new release section."
        ),
    )
    build_parser.add_argument(
        "--changelog",
        type=Path,
        help="Path to the changelog file. Defaults to '%(default)s'.",
    )
    build_parser.add_argument(
        "-V",
        "--new-version",
        required=True,
        help="The newest project version.",
    )
    build_parser.add_argument(
        "-i",
        "--in-place",
        action="store_true",
        help=(
            "Change the changelog file in-place instead of outputing the"
            " new changelog contents to STDOUT."
        ),
    )

    new_command(
        "info", help="Print internal state to standard output as JSON."
    )

    new_parser = new_command(
        "new",
        help="Add a new bullet to the KIND section of the next release.",
    )

    choices = literal_to_list(Kind)
    new_parser.add_argument(
        "kind",
        metavar="KIND",
        choices=choices,
        help=(
            "This is the type (aka KIND) of the changelog bullet that will be"
            f" added. Choose from one of {choices}."
        ),
    )
    new_parser.add_argument(
        "-b",
        "--body",
        help=(
            "The contents of the new bullet. If no body is provided,"
            " the bullet file will be opened using your system's"
            " default editor so you can provide one."
        ),
    )
    new_parser.add_argument(
        "-n",
        "--no-commit",
        dest="commit_changes",
        action="store_false",
        help=(
            "Specify this option if you do NOT want to commit this new"
            " bullet using git."
        ),
    )
    new_parser.add_argument(
        "-t",
        "--tags",
        type=clack.comma_list_or_file.parse,
        help=clack.comma_list_or_file.help(
            "Tags (e.g. a Jira issue number) to apply to the new bullet."
        ),
    )
    new_parser.add_argument(
        "-B",
        "--bullet-file-name",
        help=(
            "The basename of the bullet file which we will add this"
            " changelog bullet to. Defaults to a bullet filename of"
            " the form USER@BRANCH."
        ),
    )

    args = parser.parse_args(argv[1:])
    kwargs = clack.filter_cli_args(args)

    return kwargs
