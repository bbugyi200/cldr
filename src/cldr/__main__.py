"""Contains the main() function constructed by clack.

This function is used by setuptools as the main entry point for this program.
"""

from __future__ import annotations

import clack

from ._config import clack_parser
from ._constants import PROJECT_NAME
from ._runners import run_build, run_info, run_new


main = clack.main_factory(
    PROJECT_NAME, runners=[run_build, run_info, run_new], parser=clack_parser
)
if __name__ == "__main__":
    main()
