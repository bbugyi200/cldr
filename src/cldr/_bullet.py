"""Contains the Bullet class definition."""

from __future__ import annotations

from pathlib import Path
import re
from typing import List, Type, TypeVar, cast

from eris import ErisError, Err, Ok, Result
from pydantic.dataclasses import dataclass
from typist import PathLike, literal_to_list

from ._config import Config
from ._constants import BULLET_EXPLANATION, KIND_TO_SECTION_MAP, Kind
from ._tags import TAG_TYPES, Tag


Bullet_T = TypeVar("Bullet_T", bound="Bullet")


class BulletConfig:
    """TODO"""

    arbitrary_types_allowed = True


@dataclass(frozen=True, config=BulletConfig)
class Bullet:
    """TODO"""

    cfg: Config
    line: str
    changelog_dir: Path
    kind: Kind
    tags: List[Tag]
    body: str

    @classmethod
    def from_string(
        cls: Type["Bullet_T"],
        cfg: Config,
        line: str,
        changelog_dir: PathLike = "changelog",
    ) -> Result["Bullet_T", ErisError]:
        """TODO"""
        changelog_dir = Path(changelog_dir)

        _TAG_PATTERN = "(?:{})".format(
            "|".join(
                "(?:{})".format(tag_type.regexp) for tag_type in TAG_TYPES
            )
        )
        BULLET_PATTERN = (
            r"^[*-][ ]*(?P<kind>[a-z]+)"
            r"[ ]*(?:\((?P<tags>{0}(?:,{0})*)\))?[ ]*:"
            r"[ ]*(?P<body>.*)$"
        ).format(_TAG_PATTERN)

        if m := re.match(BULLET_PATTERN, line):
            kind = cast(Kind, m.group("kind").lower())
            if kind not in KIND_TO_SECTION_MAP:
                return Err(
                    f"An invalid bullet kind ({kind!r}) was detected in the"
                    f" following line:\n\n{line!r}\n\nUse one of the following"
                    " supported bullet types instead:"
                    f" {sorted(cast(List[Kind], literal_to_list(Kind)))}"
                )

            tags_group = m.group("tags")
            raw_tag_list = tags_group.split(",") if tags_group else []
            tags: List[Tag] = []
            for raw_tag in raw_tag_list:
                for tag_type in TAG_TYPES:
                    if re.match(tag_type.regexp, raw_tag):
                        tags.append(tag_type(raw_tag))
                        break
                else:
                    return Err(
                        "The following tag does not match any known tag"
                        f" types: {raw_tag!r}"
                    )

            return Ok(
                cls(
                    cfg,
                    line,
                    changelog_dir,
                    kind,
                    tags,
                    m.group("body"),
                )
            )
        else:
            return Err(
                f"{BULLET_EXPLANATION}\n\nThe following line does not match"
                f" the required form: {line!r}"
            )

    def to_string(self) -> str:
        """TODO"""
        result = f"* {self.body}"

        for tag in self.tags:
            result = tag.transform_bullet(self, result).unwrap()

        result += "\n"
        return result
