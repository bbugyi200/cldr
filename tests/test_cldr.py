"""Tests the cldr project's CLI."""

from __future__ import annotations

import datetime as dt
from pathlib import Path
from typing import Iterable, Sequence

from _pytest.capture import CaptureFixture
from pytest import fixture, mark, param
from syrupy.assertion import SnapshotAssertion as Snapshot

from cldr import cli


params = mark.parametrize

CONTENTS1 = """\
# ChangeLog

All notable changes to this project will be documented in this file.


## [Unreleased](https://bbgithub.dev.bloomberg.com/ComplianceSRE/tools/compare/0.9.9...HEAD)

This section's contents will just be overwritten by cldr anyway.

## [0.9.9](https://bbgithub.dev.bloomberg.com/ComplianceSRE/tools/compare/0.9.8...0.9.9) - 2021-08-08

### Fixed

* Fixed some bug.
"""

CONTENTS2 = """\
# ChangeLog

All notable changes to this project will be documented in this file.


## Unreleased
"""


@fixture(name="changelog_dir")
def changelog_dir_fixture(tmp_path: Path) -> Path:
    """TODO"""
    result = tmp_path / "changelog"
    result.mkdir()
    return result


@params(
    "contents",
    [param(CONTENTS1, id="contents1"), param(CONTENTS2, id="contents2")],
)
@params(
    "bullet,ec",
    [
        param("* add: Adding some feature.", 0, id="add-some-feature"),
        param(
            "- rm(foo-103): Removing some feature.",
            0,
            id="remove-some-feature",
        ),
        param(
            "* chg (foo-103,!123,python-libs#456): Changing some feature.",
            0,
            id="change-some-feature",
        ),
        param("* addd: Added typo...", 1, id="add-typo"),
        param("* add(ak7k2): Bad jira issue...", 1, id="bad-jira-issue"),
        param("* No bullet kind...", 1, id="no-bullet-kind"),
        param("No bullet...", 1, id="no-bullet"),
    ],
)
def test_build(
    snapshot: Snapshot,
    changelog_dir: Path,
    contents: str,
    bullet: str,
    ec: int,
) -> None:
    """TODO"""
    changelog = changelog_dir.parent / "CHANGELOG.md"
    changelog.write_text(contents)

    bullet_file = changelog_dir / "user@branch.md"
    bullet_file.write_text(bullet + "\n")

    exit_code = cli.main(
        [
            "",
            "--changelog-dir",
            str(changelog_dir),
            "build",
            "-V",
            "1.0.0",
            "--changelog",
            str(changelog),
            "-i",
        ]
    )

    # Verify that the exit code is as expected.
    assert exit_code == ec

    # Verify that the CHANGELOG.md file's contents match our expectations.
    today = dt.date.today()
    changelog_contents = changelog.read_text().replace(
        today.strftime("%Y-%m-%d"), "YYYY-MM-DD"
    )
    assert ec != 0 or changelog_contents == snapshot


@params(
    "bullets,args",
    [
        ([], ["add", "-t", "123,!5", "Added some feature."]),
        (["* add(123,!5): Added some feature."], ["fix", "Fixed some bug."]),
        (
            ["* add(123,!5): Added some feature.", "* fix: Fixed some bug."],
            ["rm", "-t", "bc", "Removed some important feature."],
        ),
    ],
)
def test_kind(
    snapshot: Snapshot,
    changelog_dir: Path,
    bullets: Iterable[str],
    args: Sequence[str],
) -> None:
    """TODO"""
    bullet_fname = "temp_bullet_file"
    bullet_file_path = changelog_dir / f"{bullet_fname}.md"
    if bullets:
        bullet_file_path.write_text(
            "\n".join(bullet for bullet in bullets) + "\n"
        )

    cmd_list = ["", "--changelog-dir", str(changelog_dir)]
    cmd_list.extend(args)
    cmd_list.extend(["--no-commit", "--bullet-file-name", bullet_fname])
    cli.main(cmd_list)

    assert bullet_file_path.read_text() == snapshot


def test_info(
    snapshot: Snapshot, capsys: CaptureFixture, changelog_dir: Path
) -> None:
    """TODO"""
    bullet_fname = "temp_bullet_file"
    bullet_file_path = changelog_dir / f"{bullet_fname}.md"
    bullet_file_path.write_text(
        "* add(123,!5): Add some new feature.\n"
        "* rm(!6,bc): Remove some old feature.\n"
    )

    cli.main(["", "--changelog-dir", str(changelog_dir), "info"])
    captured = capsys.readouterr()
    assert captured.out == snapshot
