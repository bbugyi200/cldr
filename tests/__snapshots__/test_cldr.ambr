# name: test_build[add-some-feature-contents1]
  '
  # ChangeLog
  
  All notable changes to this project will be documented in this file.
  
  
  ## [Unreleased](https://github.com/bbugyi200/cldr/compare/1.0.0...HEAD)
  
  The unreleased section is unique in that we do not add content to it directly.
  Instead, developers of this project add specially formatted bullets to files of
  the form `changelog/USER@BRANCH.md`. Refer to the [changelog/README.md] file or the
  [cldr] script (which consumes these bullets when a new version of this project
  is released) for more information.
  
  [changelog/README.md]: https://github.com/bbugyi200/cldr/tree/master/changelog
  [cldr]: https://github.com/bbugyi200/cldr
  
  
  ## [1.0.0](https://github.com/bbugyi200/cldr/compare/0.9.9...1.0.0) - YYYY-MM-DD
  
  ### Added
  
  * Adding some feature.
  
  
  ## [0.9.9](https://github.com/bbugyi200/cldr/compare/0.9.8...0.9.9) - 2021-08-08
  
  ### Fixed
  
  * Fixed some bug.
  
  '
---
# name: test_build[add-some-feature-contents2]
  '
  # ChangeLog
  
  All notable changes to this project will be documented in this file.
  
  
  ## [Unreleased](https://github.com/bbugyi200/cldr/compare/1.0.0...HEAD)
  
  The unreleased section is unique in that we do not add content to it directly.
  Instead, developers of this project add specially formatted bullets to files of
  the form `changelog/USER@BRANCH.md`. Refer to the [changelog/README.md] file or the
  [cldr] script (which consumes these bullets when a new version of this project
  is released) for more information.
  
  [changelog/README.md]: https://github.com/bbugyi200/cldr/tree/master/changelog
  [cldr]: https://github.com/bbugyi200/cldr
  
  
  ## [1.0.0](https://github.com/bbugyi200/cldr/releases/tag/1.0.0) - YYYY-MM-DD
  
  ### Added
  
  * Adding some feature.
  
  '
---
# name: test_build[change-some-feature-contents1]
  '
  # ChangeLog
  
  All notable changes to this project will be documented in this file.
  
  
  ## [Unreleased](https://github.com/bbugyi200/cldr/compare/1.0.0...HEAD)
  
  The unreleased section is unique in that we do not add content to it directly.
  Instead, developers of this project add specially formatted bullets to files of
  the form `changelog/USER@BRANCH.md`. Refer to the [changelog/README.md] file or the
  [cldr] script (which consumes these bullets when a new version of this project
  is released) for more information.
  
  [changelog/README.md]: https://github.com/bbugyi200/cldr/tree/master/changelog
  [cldr]: https://github.com/bbugyi200/cldr
  
  
  ## [1.0.0](https://github.com/bbugyi200/cldr/compare/0.9.9...1.0.0) - YYYY-MM-DD
  
  ### Changed
  
  * Changing some feature. ([FOO-103](https://jira.prod.company.com/browse/FOO-103), [PR:#123](https://github.com/bbugyi200/cldr/pull/123), [python-libs#456](https://github.com/bbugyi200/python-libs/issues/456))
  
  
  ## [0.9.9](https://github.com/bbugyi200/cldr/compare/0.9.8...0.9.9) - 2021-08-08
  
  ### Fixed
  
  * Fixed some bug.
  
  '
---
# name: test_build[change-some-feature-contents2]
  '
  # ChangeLog
  
  All notable changes to this project will be documented in this file.
  
  
  ## [Unreleased](https://github.com/bbugyi200/cldr/compare/1.0.0...HEAD)
  
  The unreleased section is unique in that we do not add content to it directly.
  Instead, developers of this project add specially formatted bullets to files of
  the form `changelog/USER@BRANCH.md`. Refer to the [changelog/README.md] file or the
  [cldr] script (which consumes these bullets when a new version of this project
  is released) for more information.
  
  [changelog/README.md]: https://github.com/bbugyi200/cldr/tree/master/changelog
  [cldr]: https://github.com/bbugyi200/cldr
  
  
  ## [1.0.0](https://github.com/bbugyi200/cldr/releases/tag/1.0.0) - YYYY-MM-DD
  
  ### Changed
  
  * Changing some feature. ([FOO-103](https://jira.prod.company.com/browse/FOO-103), [PR:#123](https://github.com/bbugyi200/cldr/pull/123), [python-libs#456](https://github.com/bbugyi200/python-libs/issues/456))
  
  '
---
# name: test_build[remove-some-feature-contents1]
  '
  # ChangeLog
  
  All notable changes to this project will be documented in this file.
  
  
  ## [Unreleased](https://github.com/bbugyi200/cldr/compare/1.0.0...HEAD)
  
  The unreleased section is unique in that we do not add content to it directly.
  Instead, developers of this project add specially formatted bullets to files of
  the form `changelog/USER@BRANCH.md`. Refer to the [changelog/README.md] file or the
  [cldr] script (which consumes these bullets when a new version of this project
  is released) for more information.
  
  [changelog/README.md]: https://github.com/bbugyi200/cldr/tree/master/changelog
  [cldr]: https://github.com/bbugyi200/cldr
  
  
  ## [1.0.0](https://github.com/bbugyi200/cldr/compare/0.9.9...1.0.0) - YYYY-MM-DD
  
  ### Removed
  
  * Removing some feature. ([FOO-103](https://jira.prod.company.com/browse/FOO-103))
  
  
  ## [0.9.9](https://github.com/bbugyi200/cldr/compare/0.9.8...0.9.9) - 2021-08-08
  
  ### Fixed
  
  * Fixed some bug.
  
  '
---
# name: test_build[remove-some-feature-contents2]
  '
  # ChangeLog
  
  All notable changes to this project will be documented in this file.
  
  
  ## [Unreleased](https://github.com/bbugyi200/cldr/compare/1.0.0...HEAD)
  
  The unreleased section is unique in that we do not add content to it directly.
  Instead, developers of this project add specially formatted bullets to files of
  the form `changelog/USER@BRANCH.md`. Refer to the [changelog/README.md] file or the
  [cldr] script (which consumes these bullets when a new version of this project
  is released) for more information.
  
  [changelog/README.md]: https://github.com/bbugyi200/cldr/tree/master/changelog
  [cldr]: https://github.com/bbugyi200/cldr
  
  
  ## [1.0.0](https://github.com/bbugyi200/cldr/releases/tag/1.0.0) - YYYY-MM-DD
  
  ### Removed
  
  * Removing some feature. ([FOO-103](https://jira.prod.company.com/browse/FOO-103))
  
  '
---
# name: test_info
  '
  {"bullets": [{"kind": "add", "body": "Add some new feature.", "tags": ["123", "!5"]}, {"kind": "rm", "body": "Remove some old feature.", "tags": ["!6", "bc"]}]}
  
  '
---
# name: test_kind[bullets0-args0]
  '
  * add(123,!5): Added some feature.
  
  '
---
# name: test_kind[bullets1-args1]
  '
  * add(123,!5): Added some feature.
  * fix: Fixed some bug.
  
  '
---
# name: test_kind[bullets2-args2]
  '
  * add(123,!5): Added some feature.
  * fix: Fixed some bug.
  * rm(bc): Removed some important feature.
  
  '
---
